File: composeApp/src/commonMain/kotlin/org/sesacteamproject/passmate/ui/home/RoomListViewModel.kt
package org.sesacteamproject.passmate.ui.home

import androidx.lifecycle.viewModelScope
import kotlinx.coroutines.Job
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch
import org.sesacteamproject.passmate.core.model.onFailure
import org.sesacteamproject.passmate.core.model.onSuccess
import org.sesacteamproject.passmate.mvi.MviViewModel
import org.sesacteamproject.passmate.payment.domain.model.RoomSort
import org.sesacteamproject.passmate.payment.domain.model.RoomTypeFilter
import org.sesacteamproject.passmate.payment.domain.usecase.GetPublicRoomsUseCase
import org.sesacteamproject.passmate.room.domain.usecase.GetRoomPinUseCase

class RoomListViewModel(
    private val getPublicRoomsUseCase: GetPublicRoomsUseCase,
    private val getRoomPinUseCase: GetRoomPinUseCase
) : MviViewModel<RoomListUiState, RoomListAction, RoomListEvent>(RoomListUiState()) {

    private var loadJob: Job? = null

    private fun onChangeQuery(query: String) {
        _uiState.update { it.copy(query = query) }
    }

    private fun onSubmitSearch() {
        reload()
    }

    private fun onSelectType(type: RoomTypeFilter) {
        if (_uiState.value.typeFilter != type) {
            _uiState.update { it.copy(typeFilter = type) }
            reload()
        }
    }

    private fun reload() {
        loadJob?.cancel()
        _uiState.update { it.copy(isLoading = true, hasError = false) }
        loadJob = viewModelScope.launch {
            val state = _uiState.value

            getPublicRoomsUseCase.invoke(
                sort = RoomSort.POPULAR,
                query = state.query.trim().ifEmpty { null },
                type = state.typeFilter,
                cursor = null
            )
                .onSuccess { page ->
                    _uiState.update {
                        it.copy(
                            isLoading = false,
                            hasError = false,
                            rooms = page.items,
                            hasNext = page.hasNext,
                            nextCursor = page.nextCursor
                        )
                    }
                }
                .onFailure {
                    _uiState.update { it.copy(isLoading = false, hasError = true) }
                }
        }
    }

    private fun onLoadMore() {
        val state = _uiState.value

        if (state.isLoading || state.isLoadingMore || !state.hasNext) {
            return
        }
        _uiState.update { it.copy(isLoadingMore = true) }
        viewModelScope.launch {
            getPublicRoomsUseCase.invoke(
                sort = RoomSort.POPULAR,
                query = state.query.trim().ifEmpty { null },
                type = state.typeFilter,
                cursor = state.nextCursor
            )
                .onSuccess { page ->
                    _uiState.update {
                        it.copy(
                            isLoadingMore = false,
                            rooms = it.rooms + page.items,
                            hasNext = page.hasNext,
                            nextCursor = page.nextCursor
                        )
                    }
                }
                .onFailure {
                    _uiState.update { it.copy(isLoadingMore = false) }
                    _event.emit(RoomListEvent.ShowNotice("목록을 더 불러오지 못했어요"))
                }
        }
    }

    // 목록 응답에 pin이 없어 roomId로 한 번 더 조회한다 (계약 `PublicRoomResponse`)
    private fun onClickRoom(roomId: Long) {
        viewModelScope.launch {
            getRoomPinUseCase.invoke(roomId)
                .onSuccess { pin -> _event.emit(RoomListEvent.OpenRoom(pin)) }
                .onFailure { _event.emit(RoomListEvent.ShowNotice("방 정보를 불러오지 못했어요. 잠시 후 다시 시도해 주세요")) }
        }
    }

    private fun onClickPinEntry() {
        viewModelScope.launch {
            _event.emit(RoomListEvent.OpenPinEntry)
        }
    }

    private fun onClickHost(hostId: Long) {
        viewModelScope.launch {
            _event.emit(RoomListEvent.OpenHostProfile(hostId))
        }
    }

    private fun onNotice(message: String) {
        viewModelScope.launch {
            _event.emit(RoomListEvent.ShowNotice(message))
        }
    }

    override fun onAction(action: RoomListAction) {
        when (action) {
            is RoomListAction.ChangeQuery -> onChangeQuery(action.query)
            is RoomListAction.SubmitSearch -> onSubmitSearch()
            is RoomListAction.SelectType -> onSelectType(action.type)
            is RoomListAction.ClickRoom -> onClickRoom(action.roomId)
            is RoomListAction.ClickHost -> onClickHost(action.hostId)
            is RoomListAction.LoadMore -> onLoadMore()
            is RoomListAction.Retry -> reload()
            is RoomListAction.ClickPinEntry -> onClickPinEntry()
            is RoomListAction.Notice -> onNotice(action.message)
        }
    }

    init {
        reload()
    }
}

File: iosApp/iosApp/ui/home/RoomListViewModel.swift
import Combine
import Foundation
import Shared

// Compose RoomListViewModel.kt 미러 — 공개 방 목록 로드·검색·필터·페이징
final class RoomListViewModel: ObservableObject {
    private let getPublicRoomsUseCase: GetPublicRoomsUseCase

    @Published private(set) var uiState = RoomListUiState()

    let event = PassthroughSubject<RoomListEvent, Never>()

    private func onChangeQuery(query: String) {
        uiState.query = query
    }

    private func onSelectType(type: RoomTypeFilter) {
        if uiState.typeFilter != type {
            uiState.typeFilter = type
            reload()
        }
    }

    private func reload() {
        uiState.isLoading = true
        uiState.hasError = false

        let query = trimmedQuery()

        getPublicRoomsUseCase.invoke(
            sort: RoomSort.popular,
            query: query,
            type: uiState.typeFilter,
            cursor: nil
        ) { [weak self] result, error in
            DispatchQueue.main.async {
                guard let self else { return }
                self.uiState.isLoading = false
                if let page = (result as? AppResultSuccess<AnyObject>)?.value as? PagedResult {
                    self.uiState.hasError = false
                    self.uiState.rooms = self.publicRooms(page)
                    self.uiState.hasNext = page.hasNext
                    self.uiState.nextCursor = page.nextCursor
                } else {
                    self.uiState.hasError = true
                }
            }
        }
    }

    private func onLoadMore() {
        if uiState.isLoading || uiState.isLoadingMore || !uiState.hasNext {
            return
        }
        uiState.isLoadingMore = true
        getPublicRoomsUseCase.invoke(
            sort: RoomSort.popular,
            query: trimmedQuery(),
            type: uiState.typeFilter,
            cursor: uiState.nextCursor
        ) { [weak self] result, error in
            DispatchQueue.main.async {
                guard let self else { return }
                self.uiState.isLoadingMore = false
                if let page = (result as? AppResultSuccess<AnyObject>)?.value as? PagedResult {
                    self.uiState.rooms += self.publicRooms(page)
                    self.uiState.hasNext = page.hasNext
                    self.uiState.nextCursor = page.nextCursor
                } else {
                    self.event.send(.showNotice(message: "목록을 더 불러오지 못했어요"))
                }
            }
        }
    }

    private func publicRooms(_ page: PagedResult) -> [PublicRoom] {
        page.items.compactMap { $0 as? PublicRoom }
    }

    private func trimmedQuery() -> String? {
        let trimmed = uiState.query.trimmingCharacters(in: .whitespaces)

        return trimmed.isEmpty ? nil : trimmed
    }

    func action(_ action: RoomListAction) {
        switch action {
        case let .changeQuery(query):
            onChangeQuery(query: query)
        case .submitSearch:
            reload()
        case let .selectType(type):
            onSelectType(type: type)
        case let .clickRoom(pin):
            event.send(.openRoom(pin: pin))
        case let .clickHost(hostId):
            event.send(.openHostProfile(hostId: hostId))
        case .loadMore:
            onLoadMore()
        case .retry:
            reload()
        case .clickPinEntry:
            event.send(.openPinEntry)
        case let .notice(message):
            event.send(.showNotice(message: message))
        }
    }

    init(getPublicRoomsUseCase: GetPublicRoomsUseCase) {
        self.getPublicRoomsUseCase = getPublicRoomsUseCase
        reload()
    }
}
